based/crates/reth/src/exec.rs

```diff
@@ -0,0 +1,438 @@
+use reth_optimism_rpc::OpReceiptBuilder;
+use reth_revm::{
+    DatabaseCommit, State,
+    context::result::{ExecutionResult, ResultAndState},
+    database::StateProviderDatabase,
+};
+use reth_rpc_convert::transaction::ConvertReceiptInput;
+use reth_storage_api::{
+    BlockReaderIdExt, BlockWriter, CanonChainTracker, DBProvider, DatabaseProviderFactory, StateProviderFactory,
+};
+use revm::database::CacheDB;
+
+use crate::{error::ExecError, unsealed_block::UnsealedBlock};
+
+const BLOCK_CACHE_LIMIT: u32 = 256;
+
+/// This trait is the ONLY place that needs to know about Reth internals.
+/// Everything else is just state-machine + bookkeeping.
+pub trait UnsealedExecutor: Send {
+    /// Ensure the executor context is ready for this env (initialize overlay state, block env, etc.)
+    fn ensure_env(&mut self, env: &EnvV0) -> Result<(), ExecError>;
+
+    /// Execute all txs in `frag` on top of current overlay state.
+    ///
+    /// MUST be cumulative: txs execute after all previous frags's txs.
+    fn execute_frag(&mut self, frag: &FragV0) -> Result<(), ExecError>;
+
+    fn seal(&mut self) -> Result<(), ExecError>;
+
+    fn set_canonical(&mut self, b: &OpBlock) -> Result<(), ExecError>;
+
+    fn get_block(&self, hash: B256) -> Result<OpBlock, ExecError>;
+
+    /// Reset overlay state completely.
+    fn reset(&mut self);
+}
+
+pub struct StateExecutor<Client> {
+    client: Client,
+    current_unsealed_block: Arc<ArcSwapOption<UnsealedBlock>>,
+    block_cache: Mutex<LruMap<B256, OpBlock>>,
+}
+
+impl<Client> StateExecutor<Client> {
+    pub fn new(client: Client) -> Self {
+        Self {
+            client,
+            current_unsealed_block: Arc::new(ArcSwapOption::new(None)),
+            block_cache: Mutex::new(LruMap::new(BLOCK_CACHE_LIMIT)),
+        }
+    }
+
+    pub fn shared_unsealed_block(&self) -> Arc<ArcSwapOption<UnsealedBlock>> {
+        Arc::clone(&self.current_unsealed_block)
+    }
+}
+
+impl<Client> UnsealedExecutor for StateExecutor<Client>
+where
+    Client: StateProviderFactory
+        + ChainSpecProvider<ChainSpec: EthChainSpec<Header = Header> + OpHardforks>
+        + BlockReaderIdExt<Header = Header, Block = OpBlock>
+        + CanonChainTracker<Header = Header>
+        + DatabaseProviderFactory
+        + Clone
+        + 'static,
+    <Client as DatabaseProviderFactory>::ProviderRW: BlockWriter<Block = OpBlock>,
+{
+    fn ensure_env(&mut self, env: &EnvV0) -> Result<(), ExecError> {
+        let Some(parent) = self.client.block_by_hash(env.parent_hash)? else {
+            return Err(ExecError::Failed(format!("parent block {} not found", env.parent_hash)))
+        };
+
+        let parent_header = parent.header();
+        let None = self.current_unsealed_block.load_full() else { return Err(ExecError::NotInitialized) };
+
+        let expected_block_number = parent_header.number.saturating_sub(1);
+        if env.number != expected_block_number {
```

Comment:
Shouldn't these validations happen with validate_env_v0 (or similar)? Why duplicate them here?

EDIT: nvm I see that this _IS_ `valida_env_v0` basically